### PR TITLE
feat(api): update user endpoints and create new endpoints for userAvatar

### PR DESCRIPTION
**Pull request by Sebastien-jo**
| Q  | A |
| --- | --- |
| Type | feat |
| Related Tickets | [PFR-53](https://project-fil-rouge.atlassian.net/browse/PFR-53) |

## What does this PR do?
Update User endpoints serialization and deserialization 
and create a new UserAvatar Entity to manage user avatar

- [ ] Unit tests
- [ ] Behat tests

[PFR-53]: https://project-fil-rouge.atlassian.net/browse/PFR-53?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/back/src/Entity/User.php [hunk 1]
```diff
@@ -16,35 +16,40 @@
...
+            uriTemplate: '/api/users/{id}/stats',
             routeName: 'api:user:stat',
-            normalizationContext: ['groups' => ['user:read']],
+            normalizationContext: ['groups' => ['user:stats']],
+            security: 'is_granted("USER_PROFILE", object)',
             name: 'api:user:stat'
         ),
-        new GetCollection(),
+        new GetCollection(
+            security: "is_granted('ADMIN')",
```
Comment (tgamiette) [inline app/back/src/Entity/User.php]: rename "ROLE_ADMIN"

File: app/back/src/Entity/User.php [hunk 2]
```diff
@@ -16,35 +16,40 @@
...
         new Post(
             denormalizationContext: ['groups' => ['user:create']],
             processor: UserPasswordHasher::class
         ),
-        new Get(uriTemplate: '/users/{id}', requirements: ['id' => '\d+']),
-        new Patch(denormalizationContext: ['groups' => ['user:update']], processor: UserPasswordHasher::class),
+        new Get(security: 'is_granted("USER_PROFILE", object)'),
+        new Patch(
+            denormalizationContext: ['groups' => ['user:update']],
+            security: 'is_granted("USER_PROFILE", object)',
```
Comment (tgamiette) [inline app/back/src/Entity/User.php]: Get  et PATCH ne sont pas sensé avoir le même voter

File: app/back/src/Entity/User.php [hunk 3]
```diff
@@ -16,35 +16,40 @@
...
+            name: 'update_user',
+            processor: UserPasswordHasher::class
+        ),
         new Delete(),
     ],
-    inputFormats: ['multipart' => ['multipart/form-data']],
     normalizationContext: ['groups' => ['user:read']],
+    denormalizationContext: ['groups' => ['user:create', 'user:update']],
```
Comment (tgamiette) [inline app/back/src/Entity/User.php]: mettre le groups de denormalization user:update dans  PATCH

File: app/back/src/Entity/User.php (hunk 1 above)
Comment (tgamiette) [reply app/back/src/Entity/User.php]: on va rename tout les roles avec le prefix ROLE_

File: app/back/src/Entity/User.php (hunk 2 above)
Comment (tgamiette) [reply app/back/src/Entity/User.php]: convention :
- USER_VIEW_LIST -> GetCollection
- USER_VIEW -> Get
- USER_CREATE -> Post 
- USER_EDIT -> Put/Patch

File: app/back/config/packages/vich_uploader.yaml [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+        type: attribute
 
     mappings:
         media_object:
             uri_prefix: /files/media
             upload_destination: '%kernel.project_dir%/public/files/medias'
             namer: Vich\UploaderBundle\Naming\SmartUniqueNamer
 
-        avatar_object:
+        imageFile:
```
Comment (tgamiette) [inline app/back/config/packages/vich_uploader.yaml]: nom trop basique
 
```suggestion
        AvatarFile:
```

File: app/back/src/Entity/Traits/TimestampableTrait.php [hunk 5]
```diff
@@ -5,14 +5,17 @@
 use DateTime;
 use DateTimeInterface;
 use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Serializer\Annotation\Groups;
 
 // class MUST implement @ORM\HasLifecycleCallbacks
 trait TimestampableTrait
 {
     #[ORM\Column(type: 'datetime')]
+    #[Groups(['user:read'])]
```
Comment (tgamiette) [inline app/back/src/Entity/Traits/TimestampableTrait.php]:
```suggestion
  
```
On va mettre un groups de normalization basique `timestamp` ( déjà dans ma PR)

File: app/back/src/Entity/Traits/TimestampableTrait.php [hunk 6]
```diff
@@ -5,14 +5,17 @@
...
     private DateTimeInterface $createdAt;
 
     #[ORM\Column(type: 'datetime')]
+    #[Groups(['user:read'])]
```
Comment (tgamiette) [inline app/back/src/Entity/Traits/TimestampableTrait.php]:
```suggestion

```

On va mettre un groups de normalization basique `timestamp` (déjà dans ma PR)

File: app/back/src/Entity/User.php (hunk 1 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/User.php]: ok :+1:

File: app/back/src/Entity/User.php (hunk 2 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/User.php]: done :heavy_check_mark:

File: app/back/src/Entity/User.php (hunk 3 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/User.php]: done :heavy_check_mark:

File: app/back/src/Entity/User.php [hunk 7]
```diff
@@ -17,35 +17,40 @@
...
+        new Patch(
+            denormalizationContext: ['groups' => ['user:update']],
+            security: 'is_granted("USER_PROFILE", object)',
+            name: 'update_user',
+            processor: UserPasswordHasher::class
+        ),
         new Delete(),
     ],
-    inputFormats: ['multipart' => ['multipart/form-data']],
     normalizationContext: ['groups' => ['user:read']],
```
Comment (tgamiette) [inline app/back/src/Entity/User.php]: rajouter le group `timestamp`

File: app/back/config/packages/vich_uploader.yaml (hunk 4 above)
Comment (Sebastien-jo) [reply app/back/config/packages/vich_uploader.yaml]: :heavy_check_mark:

File: app/back/src/Entity/Traits/TimestampableTrait.php (hunk 5 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/Traits/TimestampableTrait.php]: :heavy_check_mark:

File: app/back/src/Entity/Traits/TimestampableTrait.php (hunk 6 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/Traits/TimestampableTrait.php]: :heavy_check_mark:

File: app/back/src/Entity/User.php (hunk 7 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/User.php]: :heavy_check_mark:

File: app/back/src/Entity/UserAvatar.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+#[ORM\Entity(repositoryClass: UserAvatarRepository::class)]
+#[ORM\HasLifecycleCallbacks]
+#[ApiResource(
+    operations: [
+        new GetCollection(),
+        new Get(
+            normalizationContext: ['groups' => ['avatar:read']],
+            security: 'is_granted("USER_PROFILE", object)',
+        ),
```
Comment (tgamiette) [inline app/back/src/Entity/UserAvatar.php]: à supprimer 
on y accede via les endpoint de User garder que `post`et `delete`

File: app/back/src/Entity/UserAvatar.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        new Post(
+            inputFormats: ['multipart' => ['multipart/form-data']],
+            controller: UserAvatarController::class,
+            name: 'create_user_avatar',
+        ),
+        new Delete(
+            security: 'is_granted("USER_PROFILE", object)',
```
Comment (tgamiette) [inline app/back/src/Entity/UserAvatar.php]:
```suggestion
            security: 'is_granted("USER_AVATAR_DELETE", object)',
```

Rajouter un VOTER `ùserAvatarVoter`

Comment (tgamiette) [reply app/back/src/Entity/UserAvatar.php]: creer  USER_AVATAR_DELETE et USER_AVATAR_CREATE

File: app/back/src/Entity/UserAvatar.php (hunk 8 above)
Comment (Sebastien-jo) [reply app/back/src/Entity/UserAvatar.php]: :+1:

File: app/back/config/packages/stof_doctrine_extensions.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# Read the documentation: https://symfony.com/doc/current/bundles/StofDoctrineExtensionsBundle/index.html
+# See the official DoctrineExtensions documentation for more details: https://github.com/doctrine-extensions/DoctrineExtensions/tree/main/doc
+stof_doctrine_extensions:
+    default_locale: en_US
```
Comment (tgamiette) [inline app/back/config/packages/stof_doctrine_extensions.yaml:4]: fr ?

**Review by tgamiette — APPROVED** [review]

**Pull request merged**